### ISSUE TITLE
Typescript does not constrain the type of array elements accessed through a variable index

### ISSUE THREAD

**Issue by BalaM314** [microsoft/TypeScript#50273]
# Bug Report

Typescript does not constrain the type of array elements accessed through a variable index

### 🔎 Search Terms

variable array index array null check object possibly null

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I searched the FAQ for "array".

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.7.4#code/DYUwLgBAhgTjUE8BcEAUA7ArgWwEYhggB8IthgBKAbQF0IBeCKgVhoG4BYAKFEgEt0AExAAPBhGacu3PgDNUseAioDhIugEJGZSgG9uASEWIVQ0TQB0YAPYBlMDAEBzVBSkBfbtyA)

### 💻 Code

```ts
let array: (number | null)[] = [5];
let index = 5;

if(array[index] != null){
	array[index].toString(); //Object is possibly 'null'.
}
```

### 🙁 Actual behavior

The call to array[index].toString() gives the error "Object is possibly 'null'".

### 🙂 Expected behavior

The call to array[index].toString() should not give an error as it is inside an if block checking that array[index] is not null.

### Workaround

A non-null assertion can be used to stop the error

**Comment by MartinJohns** [comment]
Duplicate of #10530. Type narrowing does not occur for indexed access forms `e[k]` where `k` is not a literal.
